### DISCUSSION TITLE
HM-600 limit_nonpersistent_absolute über 600

### DISCUSSION THREAD

**Discussion by usekevin1988** [tbnobody/OpenDTU#406] [Q&A]
Mahlzeit!

eventuell kein Fehler, aber sicherheitshalber mal Nachfragen, vielleicht ists ja doch einer :)

Ich probiere mich gerade an einer Nulleinspeiselösung mit Node-RED.

Nun kann ich ja via MQTT  "cmd/limit_nonpersistent_absolute" einen Wert setzen. Das klappt soweit auch.

ABER, der Wert kann auch größer als 600 sein. Über die weboberfläche schmeisst er ein Error raus wenn ich mehr eingeben will. Über MQTT setzt er den Wert aber beliebig.

![grafik](https://user-images.githubusercontent.com/114338301/205491713-004ff201-0ba4-4751-b60e-82b182c1b673.png)

Ich in meinem fall habe 2x 385W Panele am WR. Macht in Summe 770W. 
Gehen wir jetzt mal vom Idealfall aus, dass die Module auch wirklich die 770W bringen. 

Wenn ich den Wert jetzt per MQTT auf 770 setze, bedeutet dies, dass da auch wirklich die 770W MAX durchgehen? oder riegelt der WR trotzdem bei 600W ab?

Weil ich glaube nicht, dass das dem WR so gut tun würde wenn man ihm mehr Watt gibt (bzw durchlässt) als er mag.

Vielleicht weiss da ja jemand mehr

**Comment by Nawile** [comment]
Hallo zusammen.

@tbnobody: Danke für dieses schhöne Projekt, dass es möglich macht sein "großes" Balkonkraftwerk selbst limitieren zu können. Ich habe damit zwei HM-600 damit regelkonform persitent auf je 300W gedrosselt.


@usekevin1988:  
Ich nutze zum ansteuern nur die Prozentangabe und über 100 wird gekappt.

Zudem habe ich testweise in Node Red ein paar Funktionen geschrieben, die beliebig viele Inverter auslesen kann und innerhalb von festlegbaren Parametern (max. Einspeisung, Bandbreite bis zum Nachregeln) die Inverter nonpersitent drosseln bzw öffnen kann.  Dazu nutze ich noch einen Hitchi IR, um de Stromzähler abzugreifen.
Gedacht ist das ganze um vielleicht zukünftig die maximal zulässige Einspeisung automatisch von der Ost- zur Westanlage  wandern zu lassen. 
In ersten Simulationen funktioniert das schon ganz gut. Jetzt ist leider seit ein paar Wochen die Sonne weg und ich hatte noch keine Lust den Ertrag zu simulieren.

Leider habe ich aktuell nicht die Zeit das richtig voran zu treiben und zu dokumentieren. Bei interesse kann ich die Nodes zum selber weiterbasteln zur Verfügung stellen.

**Reply by usekevin1988** [reply]
@Nawile 
Sehr sehr (!!!) gerne. Bei mir klappt das ganze noch nicht so, da ich noch nicht so fit bin im Umgang mit NodeRed. Der von mir beschriebene fehler ist mir dabei halt nur aufgefallen.

Ich nutze ebenfalls einen hichi IR zum Auslesen, von daher denke ich mal, dass da nur kleinere Anpassungen nötig wären damit das ganze für mich funktioniert.

**Reply by saragmx** [reply]
Guten Tag,
der **HM-600** gibt bei einer Anweisung von **1501 Watt** (number.hm_600_limit_nonpersistent_absolute) via Node Red den Hinweis aus, dass nur Werte im Bereich von **10.0 bis 1500.0 W** eingestellt werden dürfen:

###
[{"id":"7bc00982b025e12b","type":"api-call-service","z":"571620054759a80e","name":"set","server":"eb5dc721d5f8ef73","version":5,"debugenabled":true,"domain":"number","service":"set_value","areaId":[],"deviceId":["06312ff457b3ba4ea6d65bdcf05a8d84"],"entityId":["number.hm_600_limit_nonpersistent_absolute"],"data":"{\"value\":\"{{payload}}\"}","dataType":"json","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"last","x":1050,"y":360,"wires":[[]],"info":"\r\n\r\nhttps://community.home-assistant.io/t/call-service-node-data-inject-msg-payload/85573\r\n\r\n\r\nhttps://discourse.nodered.org/t/how-do-i-inject-a-service-to-a-call-service-node/46814/5\r\n\r\n\r\n\r\n\r\nsyntax spielerei: JSON\r\n\r\n{\r\n    \"entity_id\": \"test_entity\",\r\n    \"old_state\": {\r\n        \"state\": \"on\"\r\n    },\r\n    \"new_state\": {\r\n        \"state\": \"off\"\r\n    }\r\n}\r\n\r\n#######################\r\n{\"value\":{{payload}}} \r\n#########################\r\n\r\n"},{"id":"90d7542ed93bfd1b","type":"inject","z":"571620054759a80e","name":"limit abs 1501","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"number.hm_600_limit_nonpersistent_absolute","payload":"1501","payloadType":"num","x":870,"y":360,"wires":[["7bc00982b025e12b"]]},{"id":"eb5dc721d5f8ef73","type":"server","name":"Home Assistant","version":5,"addon":true,"rejectUnauthorizedCerts":true,"ha_boolean":"y|yes|true|on|home|open","connectionDelay":true,"cacheJson":true,"heartbeat":false,"heartbeatInterval":"30","areaSelector":"friendlyName","deviceSelector":"friendlyName","entitySelector":"friendlyName","statusSeparator":": ","statusYear":"hidden","statusMonth":"short","statusDay":"numeric","statusHourCycle":"default","statusTimeFormat":"h:m","enableGlobalContextStore":false}]

###

Das lässt die trügerische Vorfreude aufkommen, dass die Hardware des HM-600 evt. zu mehr fähig ist, als es die Produktbezeichung hergibt.

Meine Anfrage deshalb an Hoymiles und deren Antwort:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Hello Sandra,

Thank you for your inquiry. Please see my replies below.

1)
The microinverter won't be damaged as long as the DC voltage is within the allowable range of the microinverter.
But this might reduce the lifetime of the electronic equipment of the inside inverter if the inverter runs at full power status at all times.
Thus, This approach is not recommended by us.

2)
The max output power is 600W for HM-600 which is depends on its hardware.
No any way to set above 600W on HM-600.

---- - Fri, 30 Dec 2022 23:01:40 +0800 Sandra<xxxx@gmx.de> 写 ----

 Hello,
1. Can the inverter be damaged if the output power is continuously controlled via MQTT?

2. The HM-600 can be controlled up to 600 W output power via a graphical user interface. Via MQTT command between 10 and 1500 W. Can the HM-600 be set above 600 W without harming it?

 Thank you for a qualified response
 Sandra

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Die Interpretation der obigen Antwort überlasse ich jedem selbst.
Sandra

![Power_Limit_HM-600_01](https://user-images.githubusercontent.com/117278108/210584322-95e15835-aeb6-4ec7-8cc8-f947b5a08bd3.png)

**Reply by tbnobody** [reply]
Das Limit mit 1500 ist eine grobe Validierung die ich eingebaut habe. (Einfach weil es aktuell kein HM Modell gibt das mehr als 1,5kw kann). Durch Tests haben wir schon festgestellt, dass wenn man z.B. einen HM-800 auf 850W setzt, der Relativwert trotzdem bei 100% dann steht (Den Relativwert kann man aus dem  Inverter auslesen). Bedeutet, egal was man dem Inverter schickt, er regelt selbst auf seine max. Leistung ab.

**Reply by saragmx** [reply]
Danke fuer die Aufklaerung.

**Comment by Nawile** [comment]
Hinweis Vorab: Dies ist ein früher Entwurf für eine intelligente Invertersteuerung. Es wird keine Haftung für irgendwas übernommen. Auch ist dies eine rein technische Betrachtung und nicht eine rechtliche. Also etwas fürs Labor...Fehlerfreiheit ist auch noch nicht garantiert.

Have fun. Über Feedback freue ich mich.

Hier die Nodes, um per MQTT die DTU und die Inverter auszulesen. Jeder Inverter muss einzeln per MQTT abgefragt werden. Im Beispiel sind zwei drin.
Von "außen" kommt der aktuelle Strombezug und der Status des Hichi-Lesekopfs. Letztere hat bei mir nicht immer eine stabile WLAN-Verbindung. Daher berechne ich die Standardabweichung und wenn diese eine Weile unverändert bleibt, weiß ich, dass keine neuen Werte gesendet werden -> Das ganze als Boolean. Wenn der Lesekopf nicht sendet, schalten die Inverter auf die maximal eingestellte Einspeisung.

Es gibt die beiden Steuervariablen:

- PVConf_Einspeisung_max -> 0 wäre Nulleinspeisung
- Glaettfaktor ->  Eine Bereich in dem die Einspeisung liegen darf. Z. B.  400-600 Watt. Erst außerhalb des Bereichs werden die Inverter geregelt. Dies soll verhindern, dass die Inverter zu häufig geregelt werden müssen. (Weiß jemand, ob zu häufiges regeln den Invertern schadet?) Bei 0 gibt es keinen Bereich. Es wird permanent geregelt.

>>>EDIT- new Version below<<<

**Reply by usekevin1988** [reply]
wow, respekt. ich schau mir das die tage mal in ruhe an :) Schonmal vielen, vielen dank dafür :)

**Reply by usekevin1988** [reply]
@Nawile 

also ich hab mich heute mal versucht etwas damit zu befassen und ja, was soll ich sagen. Sieht nett aus, aber wirklich was verstehen tu ich nicht. Mir stellt sich zB die Frage wieso du das so kompliziert löst (nicht böse gemeint).

Aber im großen und ganzen wäre doch eigentlich nur folgendes nötig:

- Abfrage ob netzbezug im Minus ist (also einspeisung) (als beispiel -100)
- Wenn ja, dann abfrage wieviel W gerade produziert werden (als Beispiel 200W)

im nachfolgenden dann den Inverter per MQTT das Limit auf 100 Watt setzen.

Fertig.

Dies dann je nach belieben wiederholen


Oder hab ich da nen denkfehler?

**Reply by Nawile** [reply]
Alles ok. Ggf ist das (noch zu) kompliziert. Da kann man bestimmt demnächst noch ein Refactoring machen.
Ist wie gesagt eine Testversion, die ggf. noch zu viele "Schleifen" durch die Entwicklung drin hat. Zum Verständnis der Idee:

- Das System soll mit beliebig vielen Inverter zurechtkommen.
- Jeder Inverter soll mit nur einer MQTT-Abfrage erfasst werden.
- Eine zusätzliche Konfiguration pro Inverter soll nicht notwendig sein, daher ermittelt das Skript eigenständig die Parameter (u. a. Leistung, Auslastung, Möglichkeiten zum (Ent-)Drosseln.
- Neben dem Drosseln müssen bei höherer Nutzung die Inverter auch wieder geöffnet werden. Dabei wird versucht zu prognostizieren, welcher Inverter noch Leistung bringt (Ost/West-Problem), um auch nur die benötigte Leistung frei zu geben.
- Im Falle von Messfehlern schaltet das System auf einen sicheren 600Watt bzw xWatt Einspeisebetrieb verteilt auf n-Wechselrichter.
-  Die Wechselrichter sollen so selten wie möglich geregelt werden.

Ich nutze den Überschuss je nach Wetter zum Heizen bzw. Kühlen. Heizen ist recht Kleinteilig mit aktuell zwei (gedrosselten) Splitklima-Innengeräten, zwei 170W IR Heizungen an Tasmota-Zwischensteckern und einer 150W IR Lampe an einem Shelly-Dimmer. Damit kann ich inzwischen jedes Watt Überschuss verheizen. Das ganze schaltet in Stufen und nach Bedarf (Soll/Ist-Temperaturen) anhand des Überschusses. Daher soll mein Überschuss zwischen 300 und 600W liegen, damit das Nutzungsskript diesen verwenden kann. Somit bilden beide Regelungen eine Balance. Das Nutzungsskript läuft schon recht gut. Leider enthält es jede Menge individuelle Konfigurationsdaten. Das möchte ich mit der Invertersteuerung besser machen.

Fazit: Komplexe Nodes, aber mit nur drei Eingaben zzgl. einem MQTT-Node pro Inverter recht leicht wiederzuverwenden und noch nicht fertig ;-)

Vielleicht kannst du / jemand etwas von den Ideen verwenden.

**Reply by khmohrprivat** [reply]
> @Nawile
> 
> also ich hab mich heute mal versucht etwas damit zu befassen und ja, was soll ich sagen. Sieht nett aus, aber wirklich was verstehen tu ich nicht. Mir stellt sich zB die Frage wieso du das so kompliziert löst (nicht böse gemeint).
> 
> Aber im großen und ganzen wäre doch eigentlich nur folgendes nötig:
> 
> * Abfrage ob netzbezug im Minus ist (also einspeisung) (als beispiel -100)
> * Wenn ja, dann abfrage wieviel W gerade produziert werden (als Beispiel 200W)
> 
> im nachfolgenden dann den Inverter per MQTT das Limit auf 100 Watt setzen.
> 
> Fertig.
> 
> Dies dann je nach belieben wiederholen
> 
> Oder hab ich da nen denkfehler?

Genau so, wie du das beschreibst, werde ich es auch realisieren!

**Reply by Nawile** [reply]
Hallo, für alle die es Interessiert:

Ich habe noch etwas weiter gebaut und **diverse Bugs aus den o. g. Nodes** geholt und es läuft jetzt schon ganz zufriedenstellend.
Das Vorgehen von mir ist etwas komplexer, da ich mehrere Inverter anspreche und das ganze ggf. auch als Nulleinspeisung nutzen können möchte.
Also sind die Erweiterungen über die Basisidee von @usekevin1988 folgende Punkte:

- Mehrere Inverter ansprechen (fast beliebig viele, nur deshalb lohnt sich die komplexe Umsetzung)
- Loadbalancing und prüfen, ob ein Inverter mehr liefern könnte. Denn nur weil das Limit noch nicht 100% ist, kommt auch mehr raus, wenn das Limit erhöht wird. (Klasissche Ost/West-Verlagerung über den Tag)
- Es  muss also je nach Situation auf- bzw. abgeregelt werden
- FailSafe: Falls es zu einem Fehler kommt (z. B. Messung des Strombezug) sind die Inverter auf eine konforme Einspeisung zu setzen.
- Randbedingung: Möglichst wenige Limitänderungen am Tag.

Also bei Interesse einfach melden.

**Reply by zanycam** [reply]
Hi Nawile, hast Du die Lösung mit Node Red in Home Assistant umgesetzt und woher holst Du den Strombezug aus dem Netz (Shelly 3EM?)?
Deine Lösung sieht nach ziemlich genau dem aus was ich für mich benötige (künftig zwei HM1500).

Viele Grüße, Sandy

**Reply by strato345** [reply]
> Hallo, für alle die es Interessiert:
> 
> Ich habe noch etwas weiter gebaut und **diverse Bugs aus den o. g. Nodes** geholt und es läuft jetzt schon ganz zufriedenstellend. Das Vorgehen von mir ist etwas komplexer, da ich mehrere Inverter anspreche und das ganze ggf. auch als Nulleinspeisung nutzen können möchte. Also sind die Erweiterungen über die Basisidee von @usekevin1988 folgende Punkte:
> 
> * Mehrere Inverter ansprechen (fast beliebig viele, nur deshalb lohnt sich die komplexe Umsetzung)
> * Loadbalancing und prüfen, ob ein Inverter mehr liefern könnte. Denn nur weil das Limit noch nicht 100% ist, kommt auch mehr raus, wenn das Limit erhöht wird. (Klasissche Ost/West-Verlagerung über den Tag)
> * Es  muss also je nach Situation auf- bzw. abgeregelt werden
> * FailSafe: Falls es zu einem Fehler kommt (z. B. Messung des Strombezug) sind die Inverter auf eine konforme Einspeisung zu setzen.
> * Randbedingung: Möglichst wenige Limitänderungen am Tag.
> 
> Also bei Interesse einfach melden.

Klingt interessant. Hätte auch Interesse so etwas umzusetzen. Im Moment werkeln bei mir 3 Hoymiles mit 700/800/ 1500 Watt. Ich bin zwar kein Node red Experte aber ich traue mir zu die Vorlage auch bei mir zu implementieren.

**Reply by openKimchi** [reply]
> Aber im großen und ganzen wäre doch eigentlich nur folgendes nötig:
> 
> * Abfrage ob netzbezug im Minus ist (also einspeisung) (als beispiel -100)
> * Wenn ja, dann abfrage wieviel W gerade produziert werden (als Beispiel 200W)
> 
> im nachfolgenden dann den Inverter per MQTT das Limit auf 100 Watt setzen.
> 
> Fertig.

OK, nun hast du dein Limit auf 100 Watt gesetzt und jetzt springt ein Verbraucher an und es wird mehr Strom benötigt. Was passiert dann mit deinem Limit?

**Reply by Nawile** [reply]
@strato345 & @zanycam 
Sorry, habe ein paar Tage nicht reingeschaut.

Hier ist mein aktueller Code:
[PV-Invertersteuerung.txt](https://github.com/tbnobody/OpenDTU/files/11119790/PV-Invertersteuerung.txt)


Läuft auch inzwischen ganz gut und stabil. Das System ist auf Sicherheit ausgelegt. Das bedeutet die Inverter sind bei mir persistent auf insgesamt 600 Watt Leitung begrenzt. Die Regelung erfolgt daher flüchtig. Bei Messfehlern beim Bezug oder einem Neustart regelt das System nicht und geht in einen Sicherheitsmodus.

@zanycam: Ich habe eine Homematic CCU3 mit RaspberryMatic, Mosquitto und RedMatic. Also ein schlankes System. Den Strombezug messe ich mit einem Hichi IR an meinem Stromzähler. Zukünftig ggf. mit einem Shelly 3EM. Den muss allerdings dann eine Fachkraft installieren.

Ich freue mich über Feedback.

**Reply by strato345** [reply]
Sorry schau auch erst jetzt wieder rein. Das Thema ist ja im Moment wieder aktuell. 
@ nawile wo trage ich mein ir Lesekopf ein? Sorry kenne mich mit NOde red nicht so aus. Brauche da was Unterstützung 🙈

**Reply by FrodoVDR** [reply]
> @strato345 & @zanycam Sorry, habe ein paar Tage nicht reingeschaut.
> 
> Hier ist mein aktueller Code: [PV-Invertersteuerung.txt](https://github.com/tbnobody/OpenDTU/files/10765261/PV-Invertersteuerung.txt)
> 
> Läuft auch inzwischen ganz gut und stabil. Das System ist auf Sicherheit ausgelegt. Das bedeutet die Inverter sind bei mir persistent auf insgesamt 600 Watt Leitung begrenzt. Die Regelung erfolgt daher flüchtig. Bei Messfehlern beim Bezug oder einem Neustart regelt das System nicht und geht in einen Sicherheitsmodus.
> 
> @zanycam: Ich habe eine Homematic CCU3 mit RaspberryMatic, Mosquitto und RedMatic. Also ein schlankes System. Den Strombezug messe ich mit einem Hichi IR an meinem Stromzähler. Zukünftig ggf. mit einem Shelly 3EM. Den muss allerdings dann eine Fachkraft installieren.
> 
> Ich freue mich über Feedback.

Leider ist der Code nicht komplett, die Einspeisung des aktuellen Verbrauchs fehlt. In der Inverterregelung ist die Seriennummer eines Wechselrichters fest verdrahtet und die aktuelle Node-Red Version ist mit der Syntax in den Funktionsblöcken nicht zufrieden, es fehlte vor den msg2... das Wörtchen var.

Aber es sind einige Ansätze im Skript die ich adaptiern kann. 

Vielen Dank für die Bereitstellung, deines Codes

**Reply by strato345** [reply]
Würdest die den Code zur Verfügung stellen?

**Reply by Nawile** [reply]
@FrodoVDR: Du hast teilweise recht. Der hart verdrahtete Inverter ist ein Workaround zum setzen eines individuellen Maximalwertes und kann raus.  Danke für den "var" Hinweis (Läuft aber auch so). 
@FrodoVDR & @strato345:
Der Verbrauch fehlt allerdings nicht, sondern kommt von "außen". Bei mir ein anderer Flow. Siehe meinen Hinweis:
"Von "außen" kommt der aktuelle Strombezug und der Status des Hichi-Lesekopfs. Letztere hat bei mir nicht immer eine stabile WLAN-Verbindung. Daher berechne ich die Standardabweichung und wenn diese eine Weile unverändert bleibt, weiß ich, dass keine neuen Werte gesendet werden -> Das ganze als Boolean. Wenn der Lesekopf nicht sendet, schalten die Inverter auf die maximal eingestellte Einspeisung."

Ziel der Regelung ist die Einspeisung auf einen konstanten Wert zu halten. Daher wird der Verbrauch nur indirekt berücksichtigt.
